app/models/rich_file_manager/move.rb

```
@@ -73,21 +76,47 @@ def move_folder(region, s3)
         sanitized_new_key = new_key
       end
 
-      objects = s3.list_objects_v2(bucket: bucket_name, prefix: sanitized_old_key)
+      t1 = Time.now
+      Rails.logger.info "Moving"
+      moves = []
+
+      objects = s3.list_objects_v2(bucket: bucket_name, prefix: sanitized_old_key, encoding_type: 'url')
 
       loop do
         objects.contents.each do |object|
-          next if object.key == sanitized_old_key
-
-          new_sub_key = object.key.sub(sanitized_old_key, sanitized_new_key)
-          s3_object(s3, object.key).move_to(bucket: bucket_name, key: new_sub_key)
+          from_key = CGI::unescape(object.key)
+          to_key = from_key.sub(sanitized_old_key, sanitized_new_key)
+          multipart_copy = object.size > FIVE_MB
+          moves << [from_key, to_key, multipart_copy ]
         end
         break unless objects.next_page?
-
         objects = objects.next_page
       end
 
-      s3_object(s3, sanitized_old_key).move_to(bucket: bucket_name, key: sanitized_new_key) if s3_object(s3, sanitized_old_key).exists?
+      elapsed_ms = ((Time.now - t1)*1000).to_i
+      Rails.logger.info "Identified #{moves.count} AWS objects. Total elapsed = #{elapsed_ms}ms"
+
+      # move files and subfolders before their parent folder
+      moves.sort_by!{|move| -move[0].length}
+      queue = moves.inject(Queue.new, :push)
+      thread_count = [MAX_THREADS, moves.length].min
+
+      threads = thread_count.times.map do
+        Thread.new do
+          begin
+            while move = queue.pop(true)
+              from_key, to_key, multipart_copy = move
+              s3_object(s3, from_key).move_to(bucket: bucket_name, key: to_key, multipart_copy: multipart_copy)
+            end
+          rescue ThreadError
```

if this is a reasonable thing to do (i.e. silently suppressing an error) i think we need a comment explaining why it doesn't matter in this instance.